_posts/2017-09-24-Unit-testing-in-angular.md

```diff
@@ -0,0 +1,504 @@
+</div>
+
+# Writing tests
+## Structure
+Now that the setup is done, let’s look at how to write the tests themselves.
+First of all, test files should be named after the .ts file you're testing, but with `.spec` added to the file name (e.g. when testing `login.component.ts`, the test file should be named `login.component.spec.ts`). 
+It's best practice to keep the spec file in the same folder as the ts file. So mostly, for a component, you’ll end up with a HTML, scss, spec.ts and ts file in one folder.
+
+<p>
+    <img class="image" alt="Folder structure" src="/img/2017-09-24-unit-testing-in-angular/files.png" />
+</p>
+
+Next up, the content of a test file. 
+The jasmine spec is used format the tests.
+This means that individual tests are grouped together in a `describe` block. 
+A test itself starts with `it`. 
+Besides tests, you can also add other blocks in to a `describe`, like `beforeEach`, `beforeAll`, `afterEach`, `afterAll`... 
+What these blocks do, is quite self-explanatory. 
+Here's an example how it could be used: when testing a class, you’ll want to create an instance of that class for each test, so instead of writing the same code in each test, you could put that code in the `beforeEach` clause. 
+Simply pass `beforeEach` a function (in lambda notation) containing the code you want it to run. 
+
+Within a test itself, the class' public functions can be called and assertions can be made. 
+Assertions are made using `expect`. You can tell it what you expect the result of what you have given it to be with `toBe`, `toEqual`, `toBeTruthy`, `toBeFalsy`, `toBeNull`...
+
+Here's an example: 
+
+{% highlight coffeescript %}  
+    describe('NAME_OF_YOUR_CLASS', () => {
+        let component;
+
+        beforeEach(() => {
+            //initialize
+            component = new AppComponent();
+        });
+        
+        //Actual tests
+        it('should have a car selected', () => {
+            //assertions
+            expect(component.carSelected).toBeThruthy();
+        });
+    
+        it('should find my favorite car brand', () => {
+            //assertions
+            const carBrand = component.getFavoriteCarBrand();
+            expect(carBrand.name).toEqual('Mazda');
+        }); 
+    }); 
+{% endhighlight %}  
+
+As you can see, you can pass some text as an argument in the `describe` call. 
+This is usually the name of the class you're testing and it'll be shown when running the tests. 
+For the tests themselves, you can also pass some text which will be shown. 
+These are mainly used for you to be able to identify the tests if they fail. 
+The text should describe what's being tested, like "It should get the brand of the car", could be written as `it('should get the brand of the car', () => ...`.
+
+## Writing the actual tests
+There are two ways to write unit tests for an Angular app. 
```

Comment:
I'd say multiple ways instead of 2 ways (via constructor, ReflectiveInjector or TestBed)

Although you can use TestBed for writing unit tests, it's mostly used for integration tests (template <-> component binding, how multiple components interact with eachother, rendering, etc.)

Reply:
Didn't even think to use the ReflectiveInjector like that :D but it's kind of the same as using TestBed since you need to give it all dependencies?

Reply:
Yep very similar indeed!